torchmultimodal/modules/encoders/swin_transformer_3d_encoder.py

```diff
@@ -0,0 +1,512 @@
+        dropout (float): Dropout rate. Default: 0.0.
+        attention_dropout (float): Attention dropout rate. Default: 0.0.
+        stochastic_depth_prob: (float): Stochastic depth rate. Default: 0.0.
+        norm_layer (nn.Module): Normalization layer.  Default: nn.LayerNorm.
+    """
+
+    def __init__(
+        self,
+        dim: int,
+        num_heads: int,
+        window_size: Tuple[int, int, int] = (2, 7, 7),
+        shift_size: Tuple[int, int, int] = (0, 0, 0),
+        mlp_ratio: float = 4.0,
+        dropout: float = 0.0,
+        attention_dropout: float = 0.0,
+        stochastic_depth_prob: float = 0.0,
+        norm_layer: Callable[..., nn.Module] = nn.LayerNorm,
+    ):
+        super().__init__()
+        self.norm1 = norm_layer(dim)
+        self.attn = ShiftedWindowAttention3d(
+            dim,
+            window_size,
+            shift_size,
+            num_heads,
+            attention_dropout=attention_dropout,
+            dropout=dropout,
+        )
+        self.stochastic_depth = StochasticDepth(stochastic_depth_prob, "row")
+        self.norm2 = norm_layer(dim)
+        self.mlp = MLPBlock(dim, int(dim * mlp_ratio), dropout)
+
+    def forward(self, x: Tensor):
+        x = x + self.stochastic_depth(self.attn(self.norm1(x)))
+        x = x + self.stochastic_depth(self.mlp(self.norm2(x)))
+        return x
+
+
+# Modified from https://github.com/SwinTransformer/Video-Swin-Transformer/blob/master/mmaction/models/backbones/swin_transformer.py#L416
+class PatchEmbed3d(nn.Module):
+    """Video to Patch Embedding.
+
+    Args:
+        patch_size (Tuple[int, int, int]): Patch token size. Default: (2, 4, 4).
+        in_channels (int): Number of input channels. Default: 3
+        embed_dim (int): Number of linear projection output channels. Default: 96.
+        norm_layer (nn.Module, optional): Normalization layer. Default: None
+    """
+
+    def __init__(
+        self,
+        patch_size: Tuple[int, int, int] = (2, 4, 4),
+        in_channels: int = 3,
+        embed_dim: int = 96,
+        norm_layer: Optional[Callable[..., nn.Module]] = None,
+    ):
+        super().__init__()
+        self.patch_size = patch_size
+
+        self.proj = nn.Conv3d(
+            in_channels, embed_dim, kernel_size=patch_size, stride=patch_size
+        )
+        if norm_layer is not None:
+            self.norm = norm_layer(embed_dim)
+        else:
+            self.norm = None
+
+    def forward(self, x):
+        """Forward function."""
+        # padding
+        _, _, D, H, W = x.size()
+        pad_size = _compute_pad_size_3d((D, H, W), self.patch_size)
+        x = F.pad(x, (0, pad_size[2], 0, pad_size[1], 0, pad_size[0]))
+        x = self.proj(x)  # B C D Wh Ww
+        x = x.permute(0, 2, 3, 4, 1)  # B D Wh Ww C
+        if self.norm is not None:
+            x = self.norm(x)
+        return x
+
+
+class SwinTransformer3dEncoder(nn.Module):
```

Comment:
I agree with @datumbox  regarding the adaptation to 3d. I think it would be preferable to add it to `torchvision` directly. This main Class can be a good entry point for the unification effort, looking into it, I think we can have a common logic under `SwinTransformer` with extra parameters  in order to cater for the 2d (`SwinTransformer2d`) and 3d (`SwinTransformer3d`) versions. For BC the default values should be for the  2d version. From this main class we could drill down and look into more unification efforts (not possible in all cases probably). For example, we have `ConvNormActivationBlock` and then 2d and 3d vertions (see https://github.com/pytorch/vision/blob/49496c4f6201f05f2351788389a0863b087e78f4/torchvision/ops/misc.py#L68). This was a simpler example, but could be used as inspiration.
Let me know your thoughts @YosuaMichael. I maybe I might have overlooked some of the differences which make this even harder than I think.

Reply:
@jdsgomes I agree by add this to torchvision (planned to do later). But for having generic SwinTransformer for both 2d and 3d, I have a feeling this could be difficult, especially to make it BC. Here are some reasons I can think of (let me know if you have a good solution):
-  some of the params input differs in type. For instance the `window_size` for 2d is `int` (they assume it always a symmetry windows like 16x16), however in 3d we use `Tuple[int, int, int]` since we have windows like `(8, 7, 7)`. 
- In this 3d version, I modularize the `patch_embed` layer because the Omnivore actually use different `patch_embed` method